### ISSUE TITLE
[TRACKER] Blender FBX support bugs

### ISSUE THREAD

**Issue by RevoluPowered** [godotengine/godot#44402]
I need to fix Blender FBX support.
- [x] uv1 uv2
- [x] normals (initial investigation shows the normal data is not properly created in the FBX, workaround st->generate_normals() in master and 3.2)
- [x] materials (functions but PBR untested)
- [x] meshes (functions normally)
- [x] animation skins (FBXCluster->TransformLink; should be global bind pose as per FBX Documentation, never anything else) (report to blender, workaround with formula: mesh_global_xform * TransformLink)
- [x] key frame tracks (Time basis is broken, should use Local Start, Stop) they use the wrong convention for keyframes in the tracks (report to blender, need repro to make these messed up files, it doesn't apply to them all) 
- [ ] animations in general (blender upstream issue; Baked anims only; no support for proper tracks) (report to blender)
- [ ] bone orientation is not correctly stored in pivots (report to blender) should be stored in pre_rotation pivot bone orient
- [ ] blender doesn't export the FBX Scaling properly (report to blender)

These should be the only thing blocking exported files from Blender to be imported correctly into Maya (bi-directional FBX; without loss of data)

As you can see there are many issues which are blocking FBX from Blender, this means they need to fix the upstream bugs, adding compatibility is not a good idea here, we can do it but even I struggled to workaround these problems.

What I'll do is fork the blender plugin and workout if it's simpler to make a better fbx exporter if not I can add the compat flags into the importer. doing this though is provisional because I won't break Maya for blender support, it would have to work for both paths.

**Comment by fire** [comment]
https://blendermarket.com/products/better-fbx-importer--exporter exists so call this blender plugin something other than "Better FBX Import & Export".

**Comment by fire** [comment]
I propose that we officially block FBX versions from Blender that do not import. The QA yesterday mentioned that these FBX imports are broken. Have them fail to import rather than fail as an empty / corrupted scene.

Also document that these FBXs don't work in the newest version of 3.2 and that only XXX works. 

Document some way working around this feature compatibility breakage. I would suggest using https://blendermarket.com/products/better-fbx-importer--exporter.

@RevoluPowered will be forking the Blender FBX plugin and working on the patches needed on his own time, but I don't expect the forked Blender FBX to be done promptly.

**Comment by RevoluPowered** [comment]
> I propose that we officially block FBX versions from Blender that do not import. The QA yesterday mentioned that these FBX imports are broken. Have them fail to import rather than fail as an empty / corrupted scene.
> 
> Also document that these FBXs don't work in the newest version of 3.2 and that only XXX works.
> 
> Document some way working around this feature compatibility breakage. I would suggest using https://blendermarket.com/products/better-fbx-importer--exporter or use glTF2.
> 
> @RevoluPowered will be forking the Blender FBX plugin and working on the patches needed on his own time, but I don't expect the forked Blender FBX to be done promptly.

Anything unsupported listed above is already blocked from being imported, even Cinema4D.

3DS Max seems to work fine so left it enabled.

**Comment by RevoluPowered** [comment]
4.0 has this unblocked with a warning per a users request

**Comment by RevoluPowered** [comment]
Update:
- PRs made for initial support with some recommended settings which produce better results.
- 3.2 and master unblocked from importing fbx meshes.

**Comment by RevoluPowered** [comment]
The most major concern for these tickets in 3.2.x and 4.x is that we need to resolve:
- animation data could have an is_blender check, and then read the scene Y UP axis, and shift the animation axis for us. AFAIK this is not done normally in the FBX SDK in Maya.
- For the skin support we should have an is_blender check for it and use the old formula mesh_transform * transform_link, this is the wrong formula for Maya/Rest of the SDK, so needs to stay behind an is_blender check. 

Blender rigs should then function, but we need to really discuss with blender these problems, unity also had to hack around these bugs I'm sure. I hate putting hacks in place.

**Comment by RevoluPowered** [comment]
Update:
#46525 this fixes animation skins, normals and removes auto generated normals \o/

kenny's skin is fixed for all the characters
<img width="1133" alt="Screenshot 2021-02-28 at 18 33 26" src="https://user-images.githubusercontent.com/748770/109434775-c816cb80-7a0e-11eb-9077-721a56f0e979.png">

some other broken characters are fixed
<img width="395" alt="Screenshot 2021-02-28 at 21 17 48" src="https://user-images.githubusercontent.com/748770/109434786-d1079d00-7a0e-11eb-977a-703ecf3f00a5.png">

I made some test artwork to resolve the problem
<img width="679" alt="Screenshot 2021-02-28 at 20 59 26" src="https://user-images.githubusercontent.com/748770/109434794-d664e780-7a0e-11eb-9ed1-ea341b1b5373.png">

When I can I'll sort out the animation keyframes :)

**Comment by akien-mga** [comment]
This is probably fixed with the new FBX importer using [FBX2glTF](https://github.com/V-Sekai/FBX2glTF/releases).
You can test it in Godot 4.0 alpha 6: https://godotengine.org/article/dev-snapshot-godot-4-0-alpha-6
This new importer might also be backported for the `3.x` branch eventually.

If you still find issues with these models using the new importer, we suggest opening a new issue with details about the problems in Godot 4.0 alpha 6 or later.
